[FLINK-24859] document new connectors formats (#17839)
apache/flink: docs/content/docs/connectors/datastream/formats/parquet.md

---
title:  "Parquet files"
weight: 4
type: docs
aliases:
- /dev/connectors/formats/parquet.html
- /apis/streaming/connectors/formats/parquet.html
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

> # Parquet format

Reviewer: ```suggestion
# Parquet
```

Author: the already merged PR on "old" connector formats contain format in all the titles, I think it's better to be coherent

Reviewer: Let's fix that in a follow-up ticket then yes :)

> Flink supports reading [parquet](https://parquet.apache.org/) files and producing [Flink rows](https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/types/Row.html).

Reviewer: I'm wondering if it makes sense to include 'producing Flink rows' here. Doesn't Flink "just" support reading and writing Parquet files? Rows are just an intermediary concept of Flink itself?

Author: I added this precision in comparison with formats that generate `Pojos` or `GenericRecords` etc ...

> To use that format you need to add the dependency below to your project:

Reviewer: ```suggestion
To use the format you need to add the Flink Parquet dependency to your project.
```

Author: agree, done

```xml
{{< artifact flink-parquet >}}
```  

> The DataStream API is now compatible with both batch and streaming execution modes. 

Reviewer: This could imply that adding the dependency is necessary to make the DataStream API compatible with both batch & streaming mode :) I think we can remove this sentence.

Author: I think it is important compared to the old formats that were for DataSet only. But I agree, this sentence would be better in a common page and not in the parquet and text format pages.

> And this format is compatible with the new Source that can be used in both modes.
> Thus, you can use this format in two ways:
> - Bounded read for batch mode
> - Continuous read for streaming mode: monitors a directory for new files that appear 

Reviewer: ```suggestion
After adding the dependency, you can read and write Parquet data in either the [Batch Execution Mode or Streaming Execution Mode]({{< ref "/docs/dev/datastream/execution_mode" >}}).
```

Author: I think it is important for the user to know that for file reading in streaming mode a monitoring of the directory will be in place and the referred doc does not mention this.

**Bounded read example**:

> In this example we create a DataStream containing parquet records as flink Rows. We project the schema to read only certain fields ("f7", "f4" and "f99").  

Reviewer: ```suggestion
In this example we create a DataStream-based application containing Parquet records as Flink Rows. We project the schema to read only certain fields ("f7", "f4" and "f99").  
```

Author: actually the code snippet only creates a DataStream that reads parquet files and all necessary objects (types, format, source) but it does not create the env, the parameters, the sink etc... I think we cannot talk about an application in that case. But fixing the capitals.

> We read records by batches of 500 records. The first boolean parameter specifies if timestamp columns need to be interpreted as UTC 

Reviewer: ```suggestion
We read records in batches of 500 records. The first boolean parameter specifies if timestamp columns need to be interpreted as UTC.
```

Author: done, thanks

> and the second boolean tells the format if the projected parquet fields names are to be interpreted in a case sensitive way.

Reviewer: ```suggestion
The second boolean instructs the application if the projected Parquet fields names are to be interpreted in a case sensitive way.
```

Author: done, thanks

There is no need for a watermark strategy as records do not contain event timestamps.

```java
final LogicalType[] fieldTypes =
  new LogicalType[] {
  new DoubleType(), new IntType(), new VarCharType()
  };

final ParquetColumnarRowInputFormat<FileSourceSplit> format =
  new ParquetColumnarRowInputFormat<>(
  new Configuration(),
  RowType.of(fieldTypes, new String[] {"f7", "f4", "f99"}),
  500,
  false,
  true);
final FileSource<RowData> source =
  FileSource.forBulkFileFormat(format,  /* Flink Path */)
  .build();
final DataStream<RowData> stream =
  env.fromSource(source, WatermarkStrategy.noWatermarks(), "file-source");
```

**Continuous read example**:

> In this example we create a DataStream containing parquet records as flink Rows that will infinitely grow
> as new files are added to the directory. We monitor for new files each second.
> We project the schema to read only certain fields ("f7", "f4" and "f99").  
> We read records by batches of 500 records. The first boolean parameter specifies if timestamp columns need to be interpreted as UTC
> and the second boolean tells the format if the projected parquet fields names are to be interpreted in a case sensitive way.
> There is no need for a watermark strategy as records do not contain event timestamps.

Reviewer: ```suggestion
In this example we create a DataStream-based application containing Parquet records as Flink Rows that will infinitely grow as new files are added to the directory. We monitor for new files each second.
We project the schema to read only certain fields ("f7", "f4" and "f99").  
We read records in batches of 500 records. The first boolean parameter specifies if timestamp columns need to be interpreted as UTC, 
The second boolean instructs the application if the projected Parquet fields names are to be interpreted in a case sensitive way.
There is no need for a watermark strategy as records do not contain event timestamps.
```

Author: done, thanks
